### ISSUE TITLE
Fuzzy search often useless

### ISSUE THREAD

**Issue by richlv** [organicmaps/organicmaps#5566]
**Describe the issue**
The search seems to be fuzzy, with no option for exact search. In ~50% of observed usecases this makes search extremely limited, even useless, as search results are polluted with way too many completely unwanted results.

There are many search related issues, for example #4282, #4190 - but they are about specific locations.
Couldn't find an issue that would explicitly seek a way to replace fuzzy search with exact search.

**Steps to reproduce**
1. Search for a shop name like Rimi or Lidl while viewing Latvia.
2. Go to the map view.
3. Observe many (in some regions - most) results having nothing to do with the search query/intent.

**Expected behavior**
Shops or other entities user searches for.

**System information:**
 - Operating system and its version: [iOS 16]
 - Organic Maps version: [5.8.3]

**Additional context**
For example, a search for "Rimi" could include Rimako, Kaķu palīdzības ceentrs RMI, Galerija Rāmis, Ramirent, Riga, Remik, risinājums, RIMA, Riija, rixie.lv, Remīnes iela, Rītiņi, Rīzīmes, Rītiņu iela, Rīti, Rīmaņi, Rītišķi, Ripiņas, Rīķi and many more.
It will actually include many entities that include the city name, like police station "Rīgas reģiona pārvaldes kriminālpolicijas pārvalde".

Searching for "Lidl" is the example where the map view search results are nearly useless.
It includes things like "Lilijas", and anything with "liel" in the name.
"liel-" is a prefix for "big", and included in the name of "Lielupe", a pretty long river. And also many, many objects that are named after that river, or anything else considered bigger than something else.
"lid-" is a prefix for "fly", so anything that's got something with flying is included.
Search results also include "Eirovīns", and I'm not sure why.
"lielisks" is great. 

Even if the user could trust that only the objects with icons are relevant for their search, that would require major user education, and include things like many restaurants "Lido", restaurant "Lidojošā varde", fast food "Lielais Kristaps" and many, many more.
Even with very educated users, collision control would easily hide the actually relevant search results - "Lidl".

![I just wanted to find Lidl...](https://github.com/organicmaps/organicmaps/assets/112888/eb086744-0e85-4674-a252-a8973c11fe6e)

**Comment by vng** [comment]
@richlv I understand your pain here.
**Useful tip**: put trailing space in a search query when you need a full (not prefix) match, e.g. "lidl ", "rimi "
And the results will be 100% (almost :) relevant.

Any ideas on how to give such kind of tips to users?

I can't say for sure that we should avoid fuzzy in a viewport search. My initial idea is to make different _blue dot_ icons for fuzzy and exact matching.

**Comment by richlv** [comment]
Thanks, that's a great tip - a normal user might never discover it, though.
Such solutions seem to be way less discoverable than a button to expand search options where one option could be "exact match" :)

Adding a space definitely reduces the number of results, but still plenty of unwanted ones - RIMA, Kaķu palīdzības centrs REMI, Rīti...
Some cases are extremely confusing - a house might have an alt_name "Rīti" ("Mornings" in Latvian), and it shows up when one searches for Rimi. It looks like a completely unrelated result, until one goes to OSM data. alt_name is not shown in Organic Maps in any way, not even in object detail view.

It's not optimal for "Lidl" search either - if I zoom to central Riga and search for "lidl ", 4 out of 5 results are wrong - OM finds restaurants "Lido".
It took me a very long time to figure out why do I get a car inspection location "Tehniskās apskates stacija (TL līdz 3,5t)" when searching for Lidl. Only after failing to find it in OSM data I realised it had matched the word "līdz".

In short - trailing space is a great hint, unfortunately not a solution.
Personally, I've benefited from fuzzy search a couple of times, and it has made OM impossible to use countless times.
As much as OM team dislikes user options, there might not be a good solution without one here.

**Comment by vng** [comment]
Blue dot color for exact match and some close to grey for fuzzy?

**Comment by richlv** [comment]
That could be better - although with all this testing I recently spotted cases where shop icons were hidden due to collision control with dots.
It still leaves open the question how could all users find out (and remember) about the trailing space trick :)

**Comment by biodranik** [comment]
@vng we need consistent results in the list and in the map view. Many users won't see a map view and uninstall the app after seeing strange results in the list.

**Comment by vng** [comment]
"consistent results"?

We have a concrete problem here, please keep focused, do you have any ideas?

**Comment by biodranik** [comment]
Yes. Do not display any "fuzzy" (non-exact) search results at all.

Rare cases when user made a typo can be handled, for example, by showing a "suggestion" result "fuzzy search" (need to think about a better name). Then if user doesn't see the necessary results, pressing this "suggestion" shows additional "fuzzy" results.

**Comment by leeoniya** [comment]
this comment may be completely unconstructice, since i'm a lowly javascript dev, and cannot contribute improvements to this codebase.

however, i do have some experience with scratching my own itch of universally shitty fuzzy matching libs: https://github.com/leeoniya/uFuzzy

out of curiosity, how large are these lists of places that are in the haystack? is it 1e3, 1e4, 1e5, 1e6?

if it is "merely" 1e5 - 2e5, would it be possible to attach an example list here so i can throw it at uFuzzy and see what happens?

uFuzzy is regexp-based and someone who's motivated could potentially port its algorithm to C++ without too much effort.

**Comment by vng** [comment]
@biodranik I have doubts here (or at least should see how fuzzy suggests will look like). I'm now in Spanish-speaking country and when I can see all the exact and the misprinted results _at the same time_ (can't imagine how suggest can help here, only if all results will be like suggestions :) ) really hepls me.

I see a lot of fancy names like Fernandes - Fernandez - Hernandes - Hernandez and all of them are valid.

**Comment by OhBoy1** [comment]
Would it be possible to indicate that only exact matches should be displayed by surrounding the search target with " " or any other unused special character ? IE "Lidl". Possibly not case sensitive. I believe this would go a long way to address this.

**Comment by biodranik** [comment]
Let's check how other apps and services do search and suggestions. Our main "fuzzy" approach's issue is that we show _results_, not _suggestions_ for fuzzy/misspelled results.

Ideally, a user should see _only_ what he/she is looking for. There is no need to show tons of _something similar_. Showing _similar, but unnecessary_ results does _not_ help user to find what he/she is looking for.

So the proper approach would be to display only exact matches as results, and display misspelled "fuzzy" "results" differently, as suggestions.

**Comment by richlv** [comment]
Idea generation phase, just coming up with some - how about a small control which switches fuzzy search on/off?
I don't like hidden mechanisms like spaces much. Not discoverable, not easy to remember.
A switch would solve cases of bad matches and also cases of traveling and perhaps not knowing the correct spelling.
The switch could even default to "off", but go to "on" when there are no search results (with some warning about this, perhaps).

This might be off, but given all the different usage scenarios, it might be impossible to pick something automatic that not only would work great in most scenarios, but also be transparent in cases when it inevitably does not work great.

**Comment by richlv** [comment]
Found an older screenshot. While traveling, wanted to find some shops - Organic Maps turned out to be useless in that case. This might be improved already, but just to show the perils of fuzzy search. There were many shops with that exact name in the area.
![mercadona](https://github.com/organicmaps/organicmaps/assets/112888/d8c5ea08-664a-47b4-b79a-900e8b183d94)

**Comment by biodranik** [comment]
@richlv we are constantly improving search and results ranking. Please always use the latest app version and the latest map data to help us improve OM.
